autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/update/ActivityUpdateEvaluator.java

```diff
@@ -0,0 +1,129 @@
+
+import com.sequenceiq.cloudbreak.message.CloudbreakMessagesService;
+import com.sequenceiq.flow.api.model.FlowCheckResponse;
+import com.sequenceiq.periscope.domain.ScalingActivity;
+import com.sequenceiq.periscope.model.ScalingActivities;
+import com.sequenceiq.periscope.monitor.context.EvaluatorContext;
+import com.sequenceiq.periscope.monitor.context.ScalingActivitiesEvaluatorContext;
+import com.sequenceiq.periscope.monitor.evaluator.EvaluatorExecutor;
+import com.sequenceiq.periscope.monitor.handler.FlowCommunicator;
+import com.sequenceiq.periscope.service.ScalingActivityService;
+
+@Component("ActivityUpdateEvaluator")
+@Scope("prototype")
+public class ActivityUpdateEvaluator extends EvaluatorExecutor {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ActivityUpdateEvaluator.class);
+
+    private static final String EVALUATOR_NAME = ActivityUpdateEvaluator.class.getName();
+
+    @Inject
+    private FlowCommunicator flowCommunicator;
+
+    @Inject
+    private CloudbreakMessagesService messagesService;
+
+    @Inject
+    private ScalingActivityService scalingActivityService;
+
+    private ScalingActivities scalingActivities;
+
+    @Nonnull
+    @Override
+    public EvaluatorContext getContext() {
+        return new ScalingActivitiesEvaluatorContext(scalingActivities);
+    }
+
+    @Override
+    public String getName() {
+        return EVALUATOR_NAME;
+    }
+
+    @Override
+    public void setContext(EvaluatorContext context) {
+        this.scalingActivities = (ScalingActivities) context.getData();
+    }
+
+    @Override
+    protected void execute() {
+        if (isEmpty(scalingActivities)) {
+            LOGGER.info("No ScalingActivities found");
+            return;
+        }
+
+        LOGGER.info("Executing ScalingActivityUpdateEvaluator: {}, for {} activityIds", scalingActivities.getId(), scalingActivities.getActivityIds().size());
+        long start = now().toEpochMilli();
+        try {
+            if (scalingActivities.getActivityIds().isEmpty()) {
+                LOGGER.info("No activityIds present, skipping request to update");
+            } else {
+                Map<String, Long> activityIdsByFlowIds = scalingActivityService.findAllByIds(scalingActivities.getActivityIds()).stream()
+                        .collect(toMap(ScalingActivity::getFlowId, ScalingActivity::getId));
+                Map<Long, FlowCheckResponse> flowResponseByActivityId = flowCommunicator.getFlowStatusFromFlowIds(activityIdsByFlowIds);
+
+                Map<Long, Long> idsWithFailedFlows = flowResponseByActivityId.entrySet()
+                        .stream()
+                        .filter(e -> Boolean.TRUE.equals(e.getValue().getLatestFlowFinalizedAndFailed()) && !isNull(e.getValue().getEndTime()))
+                        .collect(toMap(Map.Entry::getKey, e -> e.getValue().getEndTime()));
+
+                List<Long> idsWithFlowsInProgress = flowResponseByActivityId.entrySet()
```

Comment:
Nit: This is where streams are not great. We're walking through the result thrice for 3 different filters, which could otherwise be a single loop - with the conditions checked, and validation to make sure we haven't left any unprocessed elements (the 3 conditions should catch every element returned)